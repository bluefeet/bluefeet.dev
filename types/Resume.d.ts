/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export type NonEmptyString = string;
export type UniqueStrings = NonEmptyString[];
export type WorkMode = "on-site" | "hybrid" | "remote";
export type EmploymentType =
  | "full-time"
  | "part-time"
  | "self-employed"
  | "freelance"
  | "contract"
  | "internship"
  | "apprenticeship"
  | "seasonal";
export type ApproximateDate = string | string;
export type Experiences = Experience[];
export type Recommendations = Recommendation[];

/**
 * The professional experiences and skills of an individual
 */
export interface Resume {
  $schema: NonEmptyString;
  contact?: Contact;
  profile?: Profile;
  objective?: Objective;
  experiences?: Experiences;
  recommendations?: Recommendations;
}
export interface Contact {
  fullName?: NonEmptyString;
  casualName?: NonEmptyString;
  pronouns?: NonEmptyString;
  location?: NonEmptyString;
  emailAddress?: string;
  phoneNumber?: string;
}
export interface Profile {
  headline?: NonEmptyString;
  about?: NonEmptyString;
  skills?: Skill[];
  languages?: UniqueStrings;
  resources?: Resource[];
}
export interface Skill {
  name: NonEmptyString;
  competencies?: UniqueStrings;
}
export interface Resource {
  title?: string;
  uri: string;
}
export interface Objective {
  /**
   * The date when first available for hire. A date in the past or no date is considered to mean immediately.
   */
  startDate?: string;
  /**
   * Taking extended time to find the best fit.
   */
  isCasual?: boolean;
  roles?: UniqueStrings;
  workModes?: WorkMode[];
  employmentTypes?: EmploymentType[];
}
export interface Experience {
  companyName?: NonEmptyString;
  title?: NonEmptyString;
  employmentType?: EmploymentType;
  location?: NonEmptyString;
  workMode?: WorkMode;
  startDate?: ApproximateDate;
  endDate?: ApproximateDate;
  summary?: NonEmptyString;
  highlights?: UniqueStrings;
}
export interface Recommendation {
  author?: Contact;
  relationship?: NonEmptyString;
  message?: NonEmptyString;
}
