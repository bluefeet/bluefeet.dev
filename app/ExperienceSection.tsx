import { Experience } from '@/types/Resume'
import Details from './Details'
import formatDateDistance from 'date-fns/formatDistance'
import headerFont from './headerFont'
import parseISODate from 'date-fns/parseISO'
import ReactMarkdown from 'react-markdown'
import resume from './resume'
import SectionTitle from './SectionTitle'
import upperFirst from 'lodash/upperFirst'
import SubSectionTitle from './SubSectionTitle'

const SingleExperience = ({ experience }: { experience: Experience }) => {
  const metaParts: string[] = []

  if (experience.title) metaParts.push(experience.title)

  if (experience.startDate) {
    const startDate = parseISODate(experience.startDate)
    const endDate = experience.endDate ? parseISODate(experience.endDate) : new Date()
    const duration = formatDateDistance(startDate, endDate)
    metaParts.push(upperFirst(duration))
  }

  if (experience.employmentType) metaParts.push(upperFirst(experience.employmentType))
  if (experience.workMode) metaParts.push(upperFirst(experience.workMode))
  if (experience.location) metaParts.push(experience.location)

  const metaLine = metaParts.join(' | ')

  return <>
    <SubSectionTitle>
      {experience.companyName}
    </SubSectionTitle>

    <div className='typography'>
      <Details>{metaLine}</Details>

      {experience.summary &&
        <ReactMarkdown>{experience.summary}</ReactMarkdown>
      }

      {experience.highlights?.length && <>
        <ul className='text-sm'>
          {experience.highlights.map((highlight) =>
            <li key={highlight}>{highlight}</li>
          )}
        </ul>
      </>}
    </div>
  </>
}

const ExperienceSection = ({ className = '' }: { className?: string }) => {
  if (!resume.experiences?.length) return <></>

  return <>
    <section className={className}>
      <SectionTitle>Experience</SectionTitle>
      {resume.experiences.map((experience) =>
        <SingleExperience
          experience={experience}
          key={`${experience.companyName}:${experience.startDate}`}
        />
      )}
    </section>
  </>
}

export default ExperienceSection
