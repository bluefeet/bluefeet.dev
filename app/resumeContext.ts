import { createContext, useContext } from 'react'
import { Resume } from '@/types/Resume'

/**
 * The default resume if none is set in the context. Exported so that tests can
 * use this as a foundation for building test resumes.
 */
export const emptyResume: Resume = {
  $schema: 'resume.schema.json',
}

const resumeContext = createContext<Resume>(emptyResume)

/**
 * Used to avoid prop drilling since the resume is accessed everywhere.
 */
export const ResumeProvider = resumeContext.Provider

/**
 * Provides access to the current resume as set by the {@link ResumeProvider}.
 */
export const useResume = () => useContext(resumeContext)
