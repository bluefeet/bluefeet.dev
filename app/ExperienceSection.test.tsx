import { render } from '@testing-library/react'

import { emptyResume, ResumeProvider } from './resumeContext'
import { Resume } from '@/types/Resume'

import { ExperienceSection } from './ExperienceSection'

const testResume = (resume: Partial<Resume>) => {
  render(
    <ResumeProvider value={{ ...emptyResume, ...resume }}>
      <ExperienceSection />
    </ResumeProvider>
  )
  expect(document.body).toMatchSnapshot()
}

describe('ExperienceSection', () => {
  it('empty resume', () => {
    testResume({})
  })

  it('multiples with required fields only', () => {
    testResume({
      experiences: [
        { companyName: 'Foo LLC' },
        { companyName: 'Bar and Daughters' },
      ]
    })
  })

  it('startDate', () => {
    testResume({
      experiences: [{
        companyName: 'Foo LLC',
        startDate: '1978-10-05'
      }]
    })
  })

  it('startDate and endDate', () => {
    testResume({
      experiences: [{
        companyName: 'Foo LLC',
        startDate: '2004-04-05',
        endDate: '2007-02-07'
      }]
    })
  })

  it('all details', () => {
    testResume({
      experiences: [{
        companyName: 'Foo LLC',
        title: 'Head Muffin Licker',
        startDate: '2004-04-05',
        endDate: '2007-02-07',
        employmentType: 'seasonal',
        workMode: 'hybrid',
      }]
    })
  })

  it('summary', () => {
    testResume({
      experiences: [{
        companyName: 'Foo LLC',
        summary: 'I did things.'
      }]
    })
  })

  it('highlights', () => {
    testResume({
      experiences: [{
        companyName: 'Foo LLC',
        highlights: [
          'Jumped.',
          'Rolled.'
        ]
      }]
    })
  })

  it('full', () => {
    testResume({
      experiences: [{
        companyName: 'Foo LLC',
        title: 'Head Muffin Licker',
        startDate: '2004-04-05',
        endDate: '2007-02-07',
        employmentType: 'seasonal',
        workMode: 'hybrid',
        summary: 'I did things.',
        highlights: [
          'Jumped.',
          'Rolled.'
        ]
      }]
    })
  })
})
