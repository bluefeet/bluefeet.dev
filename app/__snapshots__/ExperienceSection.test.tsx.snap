// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperienceSection all details 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <div
        class="text-zinc-300 text-sm mb-3"
      >
        Head Muffin Licker | Apr 2004 - Feb 2007 | Seasonal | Hybrid
      </div>
    </section>
  </div>
</body>
`;

exports[`ExperienceSection empty resume 1`] = `
<body>
  <div />
</body>
`;

exports[`ExperienceSection full 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <div
        class="text-zinc-300 text-sm mb-3"
      >
        Head Muffin Licker | Apr 2004 - Feb 2007 | Seasonal | Hybrid
      </div>
      <div
        class="typography"
      >
        I did things.
      </div>
      <div
        class="typography text-sm"
      >
        <ul>
          <li>
            Jumped.
          </li>
          <li>
            Rolled.
          </li>
        </ul>
      </div>
    </section>
  </div>
</body>
`;

exports[`ExperienceSection highlights 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <div
        class="typography text-sm"
      >
        <ul>
          <li>
            Jumped.
          </li>
          <li>
            Rolled.
          </li>
        </ul>
      </div>
    </section>
  </div>
</body>
`;

exports[`ExperienceSection multiples with required fields only 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Bar and Daughters
      </h3>
    </section>
  </div>
</body>
`;

exports[`ExperienceSection startDate 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <div
        class="text-zinc-300 text-sm mb-3"
      >
        Oct 1978 - Now
      </div>
    </section>
  </div>
</body>
`;

exports[`ExperienceSection startDate and endDate 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <div
        class="text-zinc-300 text-sm mb-3"
      >
        Apr 2004 - Feb 2007
      </div>
    </section>
  </div>
</body>
`;

exports[`ExperienceSection summary 1`] = `
<body>
  <div>
    <section
      class=""
    >
      <h2
        class="text-center print:text-left md:text-left text-sky-300 text-4xl print:text-3xl font-bold print:font-semibold mt-3 mb-4 className "
      >
        Experience
      </h2>
      <h3
        class="text-sky-400 text-3xl mt-5 mb-0 className"
      >
        Foo LLC
      </h3>
      <div
        class="typography"
      >
        I did things.
      </div>
    </section>
  </div>
</body>
`;
